File: scala/bob/src/main/scala/Bob.scala
// Fun Bob
class Bob {
  type Respond = PartialFunction[String, String]

  def silence:Respond = {case m if(m trim() isEmpty) => "Fine. Be that way!"}

  def shouting:Respond =
    {case m if(m.exists(_.isLetter) && !m.exists(_.isLower)) => "Whoa, chill out!"}

  val question:Respond ={case m if(m endsWith "?") => "Sure."}

  val whatever:Respond = {case _:String => "Whatever."}

  val hey:Respond =
    silence orElse
    shouting orElse
    question orElse
    whatever
}

File: java/bob/src/main/java/Bob.java
import java.util.stream.IntStream;

public class Bob {
    public static final String WHATEVER = "Whatever.";
    public static final String SURE = "Sure.";
    public static final String WHOA = "Whoa, chill out!";
    public static final String FINE = "Fine. Be that way!";

    public static final char last(String s) {
        return s.charAt(s.length()-1);
    }

    public static final boolean isSure(String s){
        return last(s) == '?';
    }

    public static final boolean isWhoa(String s) {
        return s.chars().noneMatch(Character::isLowerCase) &&
                s.chars().anyMatch(Character::isUpperCase);
    }

    public static final boolean isStreamEmpty(IntStream stream) {
        return !stream.findFirst().isPresent();
    }

    public static String hey(String msg) {
        final String trimmed = msg.trim();
        return trimmed.isEmpty() ? FINE :
                isWhoa(trimmed) ? WHOA :
                        isSure(trimmed) ? SURE :
                                WHATEVER;
    }
}